lib/openstack_query/api/query_api.py

```diff
@@ -24,6 +24,12 @@ def __init__(self, query_components: QueryComponents):
         self._query_results = None
         self._query_results_as_objects = None
 
+        # setup executer functions
+        self.executer.parse_func = self.parser.run_parser
+        self.executer.output_func = self.output.generate_output
```

Comment:
This is a smell, the executor should return whatever for this level to forward onto `self.parser.run_parser` rather than make them implicitly linked through the executor

The parse method call in executor should be moved up to here. It also means the executor's methods don't just become god methods in themselves

Reply:
Why are the attributes for the parser and output been linked together under `self.executer`? I'm a little confused what `self.executer` does and what it's now doing

Reply:
@meoflynn 
We need to perform two steps after generating raw results. Raw results in this case are openstack objects

1. First - group/sort them - handled by function `run_parser` 
2. Second  - convert openstack objects into dictionaries - to get only the properties we care about - handled by the function `generate_output` 

We pass a reference to these functions to the executer - so the executer can perform these steps when running `executer.parse_results` 

I did this mainly because I want the executer to be the only place to store a reference to the `raw_results`

Reply:
@DavidFair I guess I can refactor `run_parser`, and `generate_output` to be parameters for `parse_results` instead

Reply:
Alternatively, in the future we can move these to be child classes of the `executor` if you're worried about encapsulation, rather than passing methods as args

For the moment, I'm happy to take them as params just because the PR is an improvement and we can refactor in the future :+1: